netaddr.go

```diff
@@ -412,96 +323,82 @@ func (ip IP) Is6() bool {
 // returns the wrapped IPv4 address. Otherwise it returns ip, regardless
 // of its type.
 func (ip IP) Unmap() IP {
-	if !ip.Is4in6() {
-		return ip
+	if ip.Is4in6() {
+		ip.z = z4
 	}
-	a := ip.ipImpl.as16()
-	return IP{v4Addr{a[12], a[13], a[14], a[15]}}
+	return ip
 }
 
 // WithZone returns an IP that's the same as ip but with the provided
-// zone. If zone is empty, the zone is removed. If ip is an IPv4
+// zone. If zoneName is empty, the zone is removed. If ip is an IPv4
 // address it's returned unchanged.
-func (ip IP) WithZone(zone string) IP {
-	if zone == "" {
-		if z, ok := ip.ipImpl.(v6AddrZone); ok {
-			return IP{z.v6Addr}
-		}
+func (ip IP) WithZone(zoneName string) IP {
+	if !ip.Is6() {
 		return ip
 	}
-	switch ip := ip.ipImpl.(type) {
-	case v6Addr:
-		return IP{v6AddrZone{ip, zone}}
-	case v6AddrZone:
-		return IP{v6AddrZone{ip.v6Addr, zone}}
+	if zoneName == "" {
+		ip.z = z6noz
+		return ip
 	}
+	ip.z = intern.Get(zoneName)
 	return ip
 }
 
 // IsLinkLocalUnicast reports whether ip is a link-local unicast address.
 // If ip is the zero value, it will return false.
 func (ip IP) IsLinkLocalUnicast() bool {
-	// See: https://en.wikipedia.org/wiki/Link-local_address.
-	switch ip := ip.ipImpl.(type) {
-	case nil:
-		return false
-	case v4Addr:
-		return ip[0] == 169 && ip[1] == 254
-	case v6Addr:
-		return ip[0] == 0xfe && ip[1] == 0x80
-	case v6AddrZone:
-		return ip.v6Addr[0] == 0xfe && ip.v6Addr[1] == 0x80
-	default:
-		panic("netaddr: unhandled ipImpl representation")
+	if ip.Is4() {
+		return ip.a[o4+0] == 169 && ip.a[o4+1] == 254
 	}
+	if ip.Is6() {
+		return ip.a[0] == 0xfe && ip.a[1] == 0x80
+	}
+	return false
 }
 
 // IsLoopback reports whether ip is a loopback address. If ip is the zero value,
 // it will return false.
 func (ip IP) IsLoopback() bool {
-	switch ip := ip.ipImpl.(type) {
-	case nil:
-		return false
-	case v4Addr:
-		return ip[0] == 127
-	case v6Addr:
-		return string(ip[:len(v6Loopback)]) == v6Loopback
-	case v6AddrZone:
-		return string(ip.v6Addr[:len(v6Loopback)]) == v6Loopback
-	default:
-		panic("netaddr: unhandled ipImpl representation")
+	if ip.Is4() {
+		return ip.a[o4+0] == 127
 	}
+	if ip.Is6() {
+		return string(ip.a[:]) == v6Loopback
```

Comment:
If you change this to `[16]byte`, you can drop the string conversion and use == directly.

Reply:
Done.